llvm/lib/Transforms/IPO/DeadArgumentElimination.cpp

```diff
@@ -714,6 +758,83 @@ void DeadArgumentEliminationPass::PropagateLiveness(const RetOrArg &RA) {
   Uses.erase(Begin, I);
 }
 
+// Update kernel arguments table inside the integration header.
+// For example:
+//   static constexpr const bool param_omit_table[] = {
+//     // OMIT_TABLE_BEGIN
+//     // kernel_name_1
+//     false, false,     // <= update to true if the argument is dead
+//     // kernel_name_2
+//     false, false,
+//     // OMIT_TABLE_END
+//   };
+//   TODO: batch changes to multiple SYCL kernels and do one bulk update.
+constexpr StringLiteral OMIT_TABLE_BEGIN("// OMIT_TABLE_BEGIN");
+constexpr StringLiteral OMIT_TABLE_END("// OMIT_TABLE_END");
+static void updateIntegrationHeader(StringRef SyclKernelName,
+                                    const ArrayRef<bool> &ArgAlive) {
+  // TODO: Enable asserts once the driver will provide the option.
+  if (!IntegrationHeaderFileName.getNumOccurrences())
+    return;
+  // assert(IntegrationHeaderFileName.hasArgStr() &&
+  //       "Integration header file not specified!");
+  ErrorOr<std::unique_ptr<MemoryBuffer>> IntHeaderBuffer =
+      MemoryBuffer::getFile(IntegrationHeaderFileName);
+
+  if (!IntHeaderBuffer)
+    report_fatal_error("unable to read integration header file '" +
+                       IntegrationHeaderFileName +
+                       "': " + IntHeaderBuffer.getError().message());
```

Comment:
Are you going to replace this with assert(s)?
Another option to use LLVM_DEBUG and just skip the optimization.
Hard fail like this probably not the best approach for an optimization.

Reply:
I think it is a hard fail. If we did not update int header it will result in a later runtime fail.

Reply:
> If we did not update int header it will result in a later runtime fail.

If we keep all arguments, it should be okay. I mean if anything is wrong with pre-requisites, this pass can just do nothing instead of crashing and it won't break anything.

Reply:
Introduced an early exit if `IntegrationHeaderFileName` is not provided.

Reply:
Great. I think we can replace all error checking with asserts in this file.
This way we can "hard fail" on build with enabled assertions and there is no overhead on internal consistency checking on the builds w/o assertions.
```c++
if (!<cond>)
  report_fatal_error(<msg>);
```
->
```c++
assert(<cond> & <msg>);
```

I don't think we should do thorough runtime checking for integration header format. This file is auto-generated by the compiler, so it's should be validated in scope of https://github.com/intel/llvm/pull/2236. We can validate that file format is correct with asserts.

Reply:
Feel free to address this in a separate PR.